src/java/org/apache/cassandra/db/rows/TrieTombstoneMarkerImpl.java

```diff
@@ -296,4 +317,140 @@ public long unsharedHeapSize()
+        {
+            if (existing == null)
+                return this;
+
+            if (existing instanceof Covering)
+            {
+                Covering existingCovering = (Covering) existing;
+                if (!pointDeletion.supersedes(existingCovering))
+                {
+                    if (coveringDeletion == null || !coveringDeletion.supersedes(existingCovering))
+                        return null;
+                    else
+                        return coveringDeletion;
+                }
+
+                Covering newCovering = combine(coveringDeletion, existingCovering);
+                if (newCovering == coveringDeletion)
+                    return this;
+                else
+                    return new Point(pointDeletion, newCovering);
+            }
+            else if (existing instanceof Point)
+            {
+                Point existingPoint = (Point) existing;
+                Covering newCovering = combine(coveringDeletion, existingPoint.coveringDeletion);
+                Covering newPoint = combine(pointDeletion, existingPoint.pointDeletion);
+                if (newCovering == coveringDeletion && newPoint == pointDeletion)
+                    return this;
+                if (newCovering == existingPoint.coveringDeletion && newPoint == existingPoint.pointDeletion)
+                    return existingPoint;
+
+                return new Point(newPoint, newCovering);
+            }
+            else
+                throw new AssertionError("Boundaries cannot be positioned on row clusterings.");
+        }
+
+        @Override
+        public boolean hasPointData()
+        {
+            return true;
+        }
+
+        @Override
+        public TrieTombstoneMarker withUpdatedTimestamp(long l)
+        {
+            if (coveringDeletion != null)
+                return new Covering(l, coveringDeletion.localDeletionTime()); // subsumed by range deletion
+            return new Point(new Covering(l, pointDeletion.localDeletionTime()), null);
+        }
+
+        @Override
+        public boolean isBoundary()
+        {
+            return true;
+        }
+
+        @Override
+        public TrieTombstoneMarker precedingState(Direction direction)
+        {
+            return coveringDeletion;
+        }
+
+        @Override
+        public TrieTombstoneMarker restrict(boolean applicableBefore, boolean applicableAfter)
+        {
+            throw new AssertionError("Cannot have a row clustering as slice bound.");
+        }
+
+        @Override
+        public TrieTombstoneMarker asBoundary(Direction direction)
+        {
+            throw new AssertionError("Cannot have a row clustering as slice bound.");
+        }
+
+        @Override
+        public String toString()
+        {
+            return pointDeletion + (coveringDeletion != null ? "(under " + (coveringDeletion != null ? coveringDeletion : "LIVE") + ")" : "");
```

Comment:
the second coveringDeletion null-check is redundant

Reply:
Done